apps/docs/content/guides/auth/auth-hooks.mdx

```diff
@@ -268,6 +514,70 @@ revoke all
 </TabPanel>
 </Tabs>
 
+**HTTP Examples**
+
+<Tabs
+  scrollable
+  size="small"
+  type="underlined"
+  defaultActiveId="sql-limit-failed-mfa-verification-attempts"
+  queryGroup="language"
+>
+<TabPanel id="sql-limit-failed-mfa-verification-attempts" label="Limit failed MFA verification attempts">
+
+Your company requires that a user can make an incorrect MFA verification attempt no more than three times in an hour. Redis is chosen as a backing store out of convenience since it is already present in the company setup.
+
+Create a Redis database using the [Upstash Console](https://console.upstash.com/) or [Upstash CLI](https://github.com/upstash/cli) to track verification attempts.
+
+Select the Global type to minimize the latency from all edge locations. Copy the `UPSTASH_REDIS_REST_URL` and `UPSTASH_REDIS_REST_TOKEN` to your .env file. Also ensure that you have `AUTH_MFA_VERIFICATION_ATTEMPT_SECRET` set.
+
+Create a hook to read and write information from Redis, which will record the last time a user had an incorrect MFA verification attempt for a factor. For example:
+
+```sql
+import { Webhook } from "https://esm.sh/standardwebhooks@1.0.0"
+import { Redis } from 'https://deno.land/x/upstash_redis@v1.19.3/mod.ts'
+
+console.log(`Function "upstash-redis-mfa-verification-counter" up and running!`)
+
+Deno.serve(async (req) => {
+  const payload = await req.text()
+  const base64_secret = Deno.env.get('AUTH_MFA_VERIFICATION_ATTEMPT_SECRET')
+  const headers = Object.fromEntries(req.headers)
+  const wh = new Webhook(base64_secret);
+  const secondsBeforeNextRetry = 3600;
+
+
+  try {
+    const { user_id, factor_id, valid } = wh.verify(payload, headers);
+
+    const redis = new Redis({
+      url: Deno.env.get('UPSTASH_REDIS_REST_URL')!,
+      token: Deno.env.get('UPSTASH_REDIS_REST_TOKEN')!,
+    })
+
+    if (!valid) {
+      const userKey = `${user_id}:mfa_fail`
+      await redis.hincrby('mfa-attempts', userKey, 1)
+
+      const failedAttemptsCount = await redis.hget('mfa-attempts', userKey);
+      await redis.expire(userKey, secondsBeforeNextRetry)
+
+
+      if (failedAttemptsCount >= 3) {
+        return new Response(JSON.stringify({ decision: 'reject' }), { status: 403, headers: { "Content-Type": "application/json" } });
```

Comment:
I'm a bit confused here, and maybe that's because I don't understand the flow well enough.

Does `valid` denote that the MFA verification as checked by Supabase Auth failed? Then wouldn't it reject anyway on that basis alone? How does the user-supplied reject decision change the flow?

Reply:
You're right, this should be based on the timestamp not the counter. Ditto for the password verification.

Thanks for flagging!

Reply:
I'm removing the examples for now as I can't find a use case that makes sense and doesn't require reading from the database.

I would like to avoid opening a connection to the DB where possible since the Password Verification Hook is unauthenticated and it might be risky to allow access via an edge function